src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
-	if errTemplated != nil {
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, errTemplated.Error())
-		return rp, hr, errTemplated
+		return nil, err
 	}
 
-	if r.helmReleaseRequiresUpdate(ctx, hr, hrTemplate) {
-		hr.Spec = hrTemplate.Spec
-		if err = r.Client.Update(ctx, hr); err != nil {
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, err.Error())
-			return rp, hr, err
-		}
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityInfo, fmt.Sprintf("HelmRelease '%s/%s' updated", rp.Namespace, rp.GetHelmReleaseName()))
-		rp.Status.HelmRelease = rp.GetHelmReleaseName()
+	convertedObj := &unstructured.Unstructured{}
+	err = json.Unmarshal(b, convertedObj)
+	if err != nil {
+		return nil, err
 	}
 
-	return rp, hr, nil
+	return convertedObj, nil
 }
 
-func (r *RedpandaReconciler) reconcileHelmRepository(ctx context.Context, rp *v1alpha2.Redpanda) (*v1alpha2.Redpanda, *sourcev1.HelmRepository, error) {
-	// Check if HelmRepository exists or create it
-	repo := &sourcev1.HelmRepository{}
-	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: rp.Namespace, Name: rp.GetHelmRepositoryName()}, repo); err != nil {
-		if apierrors.IsNotFound(err) {
-			repo = r.createHelmRepositoryFromTemplate(rp)
-			if errCreate := r.Client.Create(ctx, repo); errCreate != nil {
-				r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("error creating HelmRepository: %s", errCreate))
-				return rp, repo, fmt.Errorf("error creating HelmRepository: %w", errCreate)
+// based on https://github.com/fluxcd/pkg/blob/30c101fc7c9fac4d84937ff4890a3da46a9db2dd/ssa/manager_apply.go#L94-L140
+func (r *RedpandaReconciler) ServerSideApply(ctx context.Context, resource *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx).WithName("RedpandaReconciler.reconcile.ServerSideApply")
+	log.V(DebugLevel).Info("resource creation", "resource", resource)
+	existingObject := &unstructured.Unstructured{}
+	existingObject.SetGroupVersionKind(resource.GetObjectKind().GroupVersionKind())
+	getError := r.Get(ctx, client.ObjectKeyFromObject(resource), existingObject)
+
+	dryRunObject := resource.DeepCopy()
+	if err := r.dryRunApply(ctx, dryRunObject); err != nil {
+		if !k8serrors.IsNotFound(getError) && isImmutableError(err) {
+			log.V(DebugLevel).Info("immutable error", "resource", resource)
+			if err := r.Delete(ctx, existingObject, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil && !k8serrors.IsNotFound(err) {
+				return err
 			}
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityInfo, fmt.Sprintf("HelmRepository '%s/%s' created ", rp.Namespace, rp.GetHelmRepositoryName()))
-		} else {
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("error getting HelmRepository: %s", err))
-			return rp, repo, fmt.Errorf("error getting HelmRepository: %w", err)
+			return r.ServerSideApply(ctx, resource)
```

Comment:
I'm not a huge fan of using recursion in cases like this as it can accidentally send us into infinite loops. Is there any reason to not just fall through to `apply` here?